eppo_core/src/event_ingestion/delivery.rs

```diff
@@ -50,108 +46,88 @@ impl DeliveryStatus {
-                        // non-retriable.
-                        delivery_status
-                            .send(DeliveryStatus::failure(batch))
-                            .await
-                            .ok()?;
-                    }
-                }
-            }
+
+        let retry_batch = std::mem::take(&mut result.retry);
+
+        let _ = delivery_status.send(result).await;
+
+        if retry_batch.is_empty() {
+            // We finished precessing the whole batch: it's either successfully delivered or failed
+            // permanently.
+            // TODO: flush
+        } else {
+            // Delay and re-insert retriable events.
+            let retries_tx = retries_tx.clone();
+            tokio::spawn(async move {
+                wait_exponential_backoff(
+                    attempts,
+                    config.max_retries,
+                    config.base_retry_delay,
+                    config.max_retry_delay,
+                )
+                .await;
+                let _ = retries_tx.send((
+                    attempts + 1,
+                    BatchedMessage {
+                        batch: retry_batch,
+                        flush,
+                    },
+                ));
+            });
         }
     }
 }
 
-fn collect_delivery_response(
-    batch: Vec<QueuedEvent>,
-    result: DeliveryResult,
+async fn wait_exponential_backoff(
+    attempts: u32,
     max_retries: u32,
-) -> DeliveryStatus {
-    if result.is_empty() {
-        return DeliveryStatus::success(batch);
-    }
-    let failed_retriable_event_uuids = result.retriable_failed_events;
-    let failed_non_retriable_event_uuids = result.non_retriable_failed_events;
-    warn!(
-        "Failed to deliver {} events (retriable)",
-        failed_retriable_event_uuids.len()
-    );
-    let mut success = Vec::new();
-    let mut failure = Vec::new();
-    let mut retry = Vec::new();
-    for queued_event in batch {
-        if failed_retriable_event_uuids.contains(&queued_event.event.uuid) {
-            if queued_event.attempts < max_retries {
-                // increment failed attempts count and retry
-                retry.push(QueuedEvent::new_from_failed(queued_event));
-            } else {
-                // max retries reached, mark as failed
-                failure.push(QueuedEvent::new_from_failed(queued_event));
-            }
-        } else if failed_non_retriable_event_uuids.contains(&queued_event.event.uuid) {
-            // event may not be retried
-            failure.push(QueuedEvent::new_from_failed(queued_event));
-        } else {
-            success.push(queued_event);
-        }
-    }
-    DeliveryStatus {
-        success,
-        failure,
-        retry,
-    }
+    min_retry_delay: Duration,
+    max_retry_delay: Duration,
+) {
+    // TODO: exponential_backoff iterator interface doesn't really suit our usage. Rewrite backoff
+    // calculation.
```

Comment:
Made `wait_exponential_backoff` infallible: if `Backoff` returns `None`, we use `max_retry_delay`. But overall, `exponential_backoff` iterator API is not really suitable for our use case, so I would rather rewrite backoff calculation or look for another library